Epsg conversion

lofrad opened the pull request:
The end2end test fails after making the changes because the output DSM doesn't have the shape expected. Need to figure out why.

carlodef committed:
[refactor] remove utm_bbx from global configuration dict

lofrad committed:
output EPSG conversion

lofrad committed:
Handle projection comment in the ply file for both UTM zone and EPSG code

lofrad committed:
Remove utm coordinates conversion in the C code, apply the conversion to out_epsg after the triangulation (for stereo only)

lofrad committed:
Handle EPSG in .ply files and for Proj objects

lofrad committed:
Change the xyz array (returned after triangulation) datatype precision from float to double

lofrad committed:
Replace iio_write_image_float_vec by iio_write_image_double_vec function

glostis commented on s2p/__init__.py:
```diff
@@ -308,10 +308,31 @@ def disparity_to_ply(tile):
+
+    # output EPSG conversion
+    in_epsg = 4979
+    if 'out_epsg' in cfg:
```
I would rather have a block like:
```python
if cfg['out_epsg'] is None:
    cfg['out_epsg'] = the local utm zone's epsg
```
somewhere in `s2p.initialisation`, and then consider everywhere downstream in the code that `cfg['out_epsg']` is defined.

What do you think about that?

glostis replied:
Also related, I would prefer to have a uniform `proj_com` in the ply files generated, rather than keeping producing some ply files with `UTM ...` and others with `EPSG ...`.
And `cfg['utm_zone']` should be dropped from everywhere in the code if possible to avoid confusion/silent inconsistencies

glostis commented on s2p/geographiclib.py:
```diff
@@ -80,25 +80,26 @@ def epsg_code_from_utm_zone(utm_zone):
 
 
-def utm_proj(utm_zone):
+def crs_proj(utm_zone, crs_type="UTM"):
```
If we drop `cfg['utm_zone']` altogether, this function could only take as input an EPSG code, and `crs_type` could be dropped.

lofrad replied:
Removing the UTM management will be more complicated as `imscript` third-party library writes the projection comment in UTM in the .ply files.

glostis replied:
Could you show me where this is done?
Because when I grep for the `crs_proj` function in the project, I only find one usage here: https://github.com/lofrad/s2p/blob/4f07fc07e70d6c50998df355d7eb731bb9591f69/s2p/rasterization.py#L126
which is then written to a `tiff` using `rasterio`, but I don't see this variable being used elsewhere.

lofrad replied:
There: https://github.com/lofrad/s2p/blob/4f07fc07e70d6c50998df355d7eb731bb9591f69/c/colormesh.c#L104
It seems to be done during the tri-stereo triangulation.
`colormesh` command is executed from here: https://github.com/lofrad/s2p/blob/4f07fc07e70d6c50998df355d7eb731bb9591f69/s2p/triangulation.py#L273
And this function is called by `heights_to_ply` in the `__init__.py` file:
https://github.com/lofrad/s2p/blob/4f07fc07e70d6c50998df355d7eb731bb9591f69/s2p/__init__.py#L464

glostis replied:
I see. `c/colormesh.c` doesn't come from `imscript` here though, it seems that its source is directly under `c/`.

It feels a bit redundant to have two pieces of code to write a point cloud to a `ply` file (one in C in `c/colormesh.c`, and the other one in Python in `s2p/ply.py`), as it requires more maintenance cost.

Ideally, the `tri-stereo` route of the code should use `s2p/ply.py` to write its point cloud, and we could get rid of `colormesh.c` altogether?
However it seems that there is a lot going on in `colormesh`, it does more than just writing a point cloud to `ply`:
- it takes a height map as input, converts it to 3D points while potentially applying a homography and clipping points to a bounding box
- there is also a `-with-normals` option. It doesn't seem to be currently used in the project so maybe it could be dropped, but I don't know what it does exactly
- the point cloud is written to a `ply`, optionally with a color image to color the points.

I guess we will need @carlodef 's advice here to decide what to do

glostis commented on s2p/initialization.py:
```diff
@@ -133,6 +134,10 @@ def build_cfg(user_cfg):
 
+    # get out_epsg
+    if 'out_epsg' not in cfg or cfg['out_epsg'] is None:
+        geographiclib.epsg_code_from_utm_zone(cfg['utm_zone'])
```
This line seems to not have any effect?

glostis reviewed (changes requested):
I have made a few comments that would require changes in several places if you agree with them, but I haven't put a comment in every single place that would require a change.

lofrad committed:
Replace xyz occurences by lonlatalt

lofrad committed:
Enable use of multiple types of proj parameters to specify the output CRS

glostis commented on s2p/initialization.py:
```diff
@@ -128,10 +129,16 @@ def build_cfg(user_cfg):
+        cfg['out_crs'] = "epsg:{}".format(epsg_code)
+    try:
+        geographiclib.pyproj_crs(cfg['out_crs'])
+    except pyproj.exceptions.CRSError as e:
```
`pyproj` is not imported in this module, so if this line is reached it raises an `ImportError`

lofrad replied:
You're right. I removed the `pyproj` import after implementing a wrapper around `pyproj.crs.CRS` in `geographiclib.py`. Addressed in 138df391d64debf207cbbc1d02fc9c3312870307

lofrad committed:
Import CRSError from pyproj.exceptions

glostis commented on s2p/geographiclib.py:
```diff
@@ -80,44 +80,51 @@ def epsg_code_from_utm_zone(utm_zone):
 
 
-def utm_proj(utm_zone):
+def crs_proj(projparams):
```
This function seems a bit redundant now that `pyproj_crs()` exists, I think we could get rid of it.

It is only used in `s2p.rasterization.plyflatten_from_plyfiles_list()`, where the `srs` attribute of the returned `Proj` object is used when writing a GeoTIFF with `rasterio`.

An easy change could be to use in `s2p.rasterization.plyflatten_from_plyfiles_list()` a `pyproj.CRS` object returned by `pyproj_crs()` instead of a `pyproj.Proj` object returned by `crs_proj()`.

The `srs`  attributes of these objects are not exactly the same, as can be seen here:
```python
In [1]: import pyproj

In [2]: pyproj.Proj("epsg:32740").srs
Out[2]: '+proj=utm +zone=40 +south +datum=WGS84 +units=m +no_defs'

In [3]: pyproj.CRS("epsg:32740").srs
Out[3]: 'epsg:32740'
```
however, I think it should work well enough with rasterio/GDAL.

If we wanted to do things really by the book, however, I think we should follow what is said in the pyproj documentation's section [Converting from pyproj.crs.CRS to rasterio.crs.CRS](https://pyproj4.github.io/pyproj/stable/crs_compatibility.html#converting-from-pyproj-crs-crs-to-rasterio-crs-crs):
```python
from distutils.version import LooseVersion

import rasterio
from rasterio.crs import CRS as RioCRS
from pyproj.crs import CRS
from pyproj.enums import WktVersion

proj_crs = CRS.from_epsg(4326)
if LooseVersion(rasterio.__gdal_version__) < LooseVersion("3.0.0")
    rio_crs = RioCRS.from_wkt(proj_crs.to_wkt(WktVersion.WKT1_GDAL))
else:
    rio_crs = RioCRS.from_wkt(proj_crs.to_wkt())
```

lofrad replied:
I took your comment into account and proposed this implementation: 43cdbc62fd8642fd95770d724b0abf9a8bc40720

lofrad committed:
Return rasterio.CRS object instead of PROJ instance

lofrad committed:
Change projection comment in input test data cloud.ply

carlodef committed:
Merge branch 'master' into remove_utm_bbx

glostis committed:
Fix failing test_distributed_plyflatten

lofrad committed:
Update docstring

glostis commented on s2p/config.py:
```diff
@@ -156,3 +156,6 @@
 # clean height maps outliers
 cfg['cargarse_basura'] = True
+
+# output coordinate reference system
```
Maybe we could document a bit examples of supported formats:

```suggestion
# Output coordinate reference system
# All formats accepted by `pyproj.CRS()` are allowed, for example:
# 32740 (int interpreted as an EPSG code), or
# "epsg:32740+5773" (authority string), or
# "+proj=utm +zone=40 +south +datum=WGS84 +units=m +vunits=m +no_defs +type=crs" (proj4 string)
# If None, the local UTM zone will be used
```

glostis commented on s2p/geographiclib.py:
```diff
@@ -5,8 +5,13 @@
 import pyproj
+import rasterio
+from rasterio.crs import CRS as RioCRS
+from pyproj.crs import CRS
```
Unused import

```suggestion
```

glostis commented on s2p/geographiclib.py:
```diff
@@ -80,44 +85,56 @@ def epsg_code_from_utm_zone(utm_zone):
 
     Returns:
-        pyproj.Proj: object that can be used to transform coordinates
+        rasterio.CRS: object that can be used with rasterio
```
Typo fix

```suggestion
        rasterio.crs.CRS: object that can be used with rasterio
```

glostis commented on s2p/geographiclib.py:
```diff
@@ -80,44 +85,56 @@ def epsg_code_from_utm_zone(utm_zone):
+def rasterio_crs(projparams):
     """
-    Return a pyproj.Proj object that corresponds to the given utm_zone string.
+    Return a rasterio.CRS object that corresponds to the given parameters.
```
Typo fix

```suggestion
    Return a rasterio.crs.CRS object that corresponds to the given parameters.
```

carlodef committed:
[tests] update expected output of rasterization.plyflatten_from_plyfiles_list

The definition of the DSM sampling grid computed by this function has
been updated so that its vertices are integer multiples of the grid cell
size. Hence the output DSM is slightly different and the expected test
result must be updated accordingly.

carlodef committed:
[refactor] remove dead "if" block

carlodef committed:
[bugfix] add missing import

carlodef committed:
[refactor] remove useless read/write

carlodef committed:
[dev] activate compression of output DSMs

carlodef committed:
[refactor] remove dead code branch

As the pointing correction step doesn't involve anymore calls to
subprocess.run, it can't raise exceptions CalledProcessError and
TimeoutExpired

carlodef committed:
[refactor] remove unused variables

carlodef committed:
[refactor] remove dead code in  visualisation.py

carlodef commented on s2p/__init__.py:
```diff
@@ -305,13 +305,33 @@ def disparity_to_ply(tile):
                                                img_bbx=(x, x+w, y, y+h),
                                                A=np.loadtxt(pointing))
 
+    # output CRS conversion
+    in_crs = geographiclib.pyproj_crs("epsg:4979")
+    pyproj_out_crs = geographiclib.pyproj_crs(cfg['out_crs'])
+    proj_com = "CRS {}".format(cfg['out_crs'])
+
+    if pyproj_out_crs != in_crs:
+
+        # reshape the xyz array into a 3-column 2D-array
+        lonlatalt_shape = lonlatalt_array.shape
+        lonlatalt_array = lonlatalt_array.reshape(-1, 3)
+        xyz_array = np.empty(lonlatalt_shape, dtype=np.float32)
+
+        x, y, z = geographiclib.pyproj_transform(lonlatalt_array[:,0], lonlatalt_array[:,1],
+                                             in_crs, pyproj_out_crs, lonlatalt_array[:,2])
+
+        xyz_array[:,:,0] = x.reshape(*lonlatalt_shape[:2])
+        xyz_array[:,:,1] = y.reshape(*lonlatalt_shape[:2])
+        xyz_array[:,:,2] = z.reshape(*lonlatalt_shape[:2])
+    else:
+        xyz_array = lonlatalt_array
+
```
Hello @lofrad and @glostis, I would suggest keeping the name `triangulation.disp_to_xyz` and moving this block inside that function

lofrad replied:
Thank you Carlo, addressed in a501bb8a522c6fe5fd78e00ce6cd89eff0e52efb

carlodef commented on c/disp_to_h.c:
```diff
@@ -11,8 +11,6 @@
 #include "pickopt.c"
 
 
-void utm_alt_zone(double *out, double lat, double lon, int zone);
-
```
As you're removing the dependency to geographiclib, you could update the `disp_to_h` compilation rule in the `makefile` to make it simpler.

lofrad replied:
Edited in f64d661d8612b2b54fc6fc1f009a963eda2245db

lofrad committed:
Rewrite heigth map to point clound C function in Python

lofrad committed:
Fix tests by removing conflicts with the out_crs key from s2p config

lofrad committed:
Move out CRS conversion to disp_to_xyz function in triangulation

lofrad committed:
Update makefile, remove geographiclib dependency for disp_to_h shared library

lofrad committed:
Apply suggestions from code review (Guillaume's comments)

Co-authored-by: Guillaume Lostis <g.lostis@kayrros.com>

lofrad committed:
Change variable name

Co-authored-by: Guillaume Lostis <g.lostis@kayrros.com>

glostis commented:
This PR was just a draft to prepare the branch for a PR on the official `cmla/s2p` repository.

Closing in favor of https://github.com/cmla/s2p/pull/53
